### PR TITLE
Refresh roles, auth fix

### LINKED ISSUE #2: CVUT Verification
https://auth.fit.cvut.cz/manager

### PR DESCRIPTION
**Pull request by tenhobi**
This proposal should fix authorization for different users:

If a new user appears:
- fetch from usermap
- grant roles

If valid user appears (both that is on a server and that rejoined the server):
- ungrant all roles from that pool
- grants roles from pool according to usermap

If user that has mismatched verification from discord id and auth hash appears:
- error

### CODE REVIEW

**Comment by tenhobi** [comment]
Tested that (cannot test staff tho), seems all right.

Tested:

- valid discord & auth
- invalid discord & valid auth
- valid discord & invalid auth
- invalid discord & invalid auth

What is the best way to send some (error only?) response to Discord user from that web callback?

File: src/HonzaBotner.Discord.Services/Commands/AuthorizeCommands.cs [hunk 1]
```diff
@@ -27,13 +27,14 @@ public async Task AuthorizeCommand(CommandContext ctx)
             DiscordUser user = message.Author;
             DiscordDmChannel channel = await message.Channel.Guild.Members[user.Id].CreateDmChannelAsync();
 
+            string link = _urlProvider.GetAuthLink(user.Id, RolesPool.Auth);
+
             if (await _authorizationService.IsUserVerified(user.Id))
             {
-                await channel.SendMessageAsync($"You are already authorized");
+                await channel.SendMessageAsync($"You are already authorized.\nTo update roles follow this link: {link}");
```
Comment (ostorc) [inline src/HonzaBotner.Discord.Services/Commands/AuthorizeCommands.cs:34]: All messages to user should be in Czech (not including commands description)

File: src/HonzaBotner.Services.Contract/IAuthorizeService.cs [hunk 2]
```diff
@@ -1,12 +1,21 @@
...
+using System.ComponentModel;
 using System.Threading.Tasks;
 using HonzaBotner.Services.Contract.Dto;
 
 namespace HonzaBotner.Services.Contract
 {
     public interface IAuthorizationService
     {
-        Task<bool> AuthorizeAsync(string accessToken, string username, ulong userId, RolesPool rolesPool);
+        public enum AuthorizeResult
```
Comment (ostorc) [inline src/HonzaBotner.Services.Contract/IAuthorizeService.cs:10]: Consider extracting this enum from interface.

File: src/HonzaBotner.Services.Contract/IDiscordRoleManager.cs [hunk 3]
```diff
@@ -9,5 +9,7 @@ public interface IDiscordRoleManager
         HashSet<DiscordRole> MapUsermapRoles(IReadOnlyCollection<string> kosRoles, RolesPool rolesPool);
 
         Task<bool> GrantRolesAsync(ulong userId, IReadOnlySet<DiscordRole> discordRoles);
+
+        Task<bool> UngrantRolesPoolAsync(ulong userId, RolesPool rolesPool);
```
Comment (ostorc) [inline src/HonzaBotner.Services.Contract/IDiscordRoleManager.cs:13]: Why Grant uses discordRoles and this one uses pool?

File: src/HonzaBotner.Services/CvutAuthorizationService.cs [hunk 4]
```diff
@@ -40,42 +40,64 @@ public class CvutAuthorizationService : IAuthorizationService
...
             IReadOnlySet<DiscordRole> discordRoles = _roleManager.MapUsermapRoles(person.Roles, rolesPool);
-            bool rolesGranted = await _roleManager.GrantRolesAsync(userId, discordRoles);
 
-            if (rolesGranted && requiresUnverified)
+            // discord and auth -> update roles
+            if (discordIdPresent && authPresent)
             {
-                Verification verification = new() {AuthId = authId, UserId = userId};
+                Verification discordVerification = await _dbContext.Verifications.FirstAsync(v => v.UserId == userId);
+                Verification authVerification = await _dbContext.Verifications.FirstAsync(v => v.AuthId == authId);
```
Comment (ostorc) [inline src/HonzaBotner.Services/CvutAuthorizationService.cs]: We can use Any to look into database. So maybe use something like `Verifications.Any(v => v.AuthId == authId && v.UserId == userId)`

File: src/HonzaBotner.Services/CvutAuthorizationService.cs [hunk 5]
```diff
@@ -40,42 +40,64 @@ public class CvutAuthorizationService : IAuthorizationService
...
+
+                if (discordVerification.Equals(authVerification))
+                {
+                    bool ungranted = await _roleManager.UngrantRolesPoolAsync(userId, rolesPool);
+                    if (!ungranted)
+                    {
+                        // TODO: error ungranting roles
```
Comment (ostorc) [inline src/HonzaBotner.Services/CvutAuthorizationService.cs]: Log error message.

File: src/HonzaBotner.Services/DiscordRoleManager.cs [hunk 6]
```diff
@@ -53,6 +53,39 @@ public async Task<bool> GrantRolesAsync(ulong userId, IReadOnlySet<DiscordRole>
...
+            };
+
+            List<DRole> roles = new();
+
+            foreach ((var key, ulong value) in rolesMapping)
+            {
+                DRole? role = guild.GetRole(value);
+                if (role == null)
+                {
+                    return false;
```
Comment (ostorc) [inline src/HonzaBotner.Services/DiscordRoleManager.cs]: Use soft approach. We want to remove as much as possible. Also log warning.

File: src/HonzaBotner/Controllers/AuthController.cs [hunk 7]
```diff
@@ -66,17 +65,28 @@ public async Task<ActionResult> Callback()
...
+            return Ok(await _authorizationService.AuthorizeAsync(accessToken, userName, userId, rolesPool) switch
+            {
+                IAuthorizationService.AuthorizeResult.OK => "Sucessfully authenticated.",
+                IAuthorizationService.AuthorizeResult.Failed => "Authentication failed.",
+                IAuthorizationService.AuthorizeResult.DifferentMember =>
+                    "Authentication failed because you are registered with another auth code or another user already uses your auth code.",
+                IAuthorizationService.AuthorizeResult.UserMapError =>
+                    "Authentication failed due to UserMap service failure.",
+                _ => throw new ArgumentOutOfRangeException()
+            });
```
Comment (ostorc) [inline src/HonzaBotner/Controllers/AuthController.cs]: Much better than previous state. In future we would ideally want some html page.

File: src/HonzaBotner/Controllers/AuthController.cs [hunk 8]
```diff
@@ -66,17 +65,28 @@ public async Task<ActionResult> Callback()
                 return Content(e.Message, "text/html");
             }
 
-            bool auth = await _authorizationService.AuthorizeAsync(accessToken, userName, userId, rolesPool);
-
-            return Ok(auth);
+            return Ok(await _authorizationService.AuthorizeAsync(accessToken, userName, userId, rolesPool) switch
```
Comment (ostorc) [inline src/HonzaBotner/Controllers/AuthController.cs]: Ideally we would want to catch possible exceptions from this call.

**Review by ostorc — CHANGES_REQUESTED** [review]

**Pull request merged**